api/controllers/AssignmentController.js

```diff
@@ -79,38 +77,19 @@ module.exports = class AssignmentController extends classMixin(OrderedDataContro
       private: parameterizedAssignment.dataValues.auxPrivate
     }
 
+    return this.processSolution(req, res, aux, assignment, solution, parameterizedAssignment)
+  }
+
+  async processSyncJudge (req, res, aux, assignment, solution, parameterizedAssignment) {
     let judgement = { grade: 0, msg: 'Wrong! Please try again.' }
+    const judge = new assignment.Judge(assignment.judge, req.user)
 
     try {
       judgement = await judge.judge(aux, req.user, assignment.Judge, solution)
-      const [solutionModel] = await Solution.findOrCreate(
-        {
-          where: { studentId: req.user.id, parameterizedAssignmentId: paramId },
-          defaults: {
-            studentId: req.user.id,
-            parameterizedAssignmentId: paramId
-          }
-        }
-      )
-
-      if (!parameterizedAssignment.dataValues.solved) {
-        await solutionModel.update({ data: solution })
-      }
+      await this.updateSolution(req, parameterizedAssignment, solution)
+      await this.updateSolved(judgement, parameterizedAssignment)
```

Comment:
Should these be in the try/catch?

Reply:
They are, don't they ?

Reply:
Yes. Why?

Reply:
Because we update database entries and we want to catch db errors. It is wrong to do it there ?

Reply:
I'm thinking that maybe we want to separate errors caused by the user (which should be reported back to them) from errors caused by us (e.g. db failures). Of course both should raise exceptions, but maybe we want to at least report them differently to the user. For example, in case it's our own error, we can say something like "Server error: (error message here)" so that the student is not confused thinking it's his own fault.

Reply:
I agree. You are right.

Reply:
Please fix this and the rest of the outstanding issues before merging.